File: src/renderer/src/services/ipc/sftp/transfer.ts
import { backendService } from '../backend'
import { TransferProgress, IPCMessage } from '../../../types'

export const upload = (
  sessionId: string,
  localPath: string,
  remotePath: string,
  onProgress?: (progress: TransferProgress) => void
): Promise<void> => {
  return new Promise((resolve, reject) => {
    let requestId = ''
    let progressHandler: ((message: IPCMessage) => void) | undefined

    const cleanup = (handler: (message: IPCMessage) => void) => {
      backendService.off('sftp:upload', handler)
      backendService.off('error', handler)
      if (progressHandler) {
        backendService.off('sftp:progress', progressHandler)
      }
    }

    if (onProgress) {
      progressHandler = (message: IPCMessage) => {
        if (
          message.request_id === requestId &&
          message.session_id === sessionId &&
          message.type === 'sftp:progress'
        ) {
          onProgress(message.data as TransferProgress)
        }
      }
      backendService.on('sftp:progress', progressHandler)
    }

    const handler = (message: IPCMessage) => {
      if (message.request_id !== requestId) return

      if (message.session_id === sessionId && message.type === 'sftp:upload') {
        cleanup(handler)
        resolve()
      } else if (message.type === 'error') {
        cleanup(handler)
        reject(new Error(message.data.error))
      }
    }

    backendService.on('sftp:upload', handler)
    backendService.on('error', handler)

    requestId = backendService.send({
      type: 'sftp:upload',
      session_id: sessionId,
      data: { local_path: localPath, remote_path: remotePath }
    })
  })
}

export const download = (
  sessionId: string,
  remotePath: string,
  localPath: string,
  onProgress?: (progress: TransferProgress) => void
): Promise<void> => {
  return new Promise((resolve, reject) => {
    let requestId = ''
    let progressHandler: ((message: IPCMessage) => void) | undefined

    const cleanup = (handler: (message: IPCMessage) => void) => {
      backendService.off('sftp:download', handler)
      backendService.off('error', handler)
      if (progressHandler) {
        backendService.off('sftp:progress', progressHandler)
      }
    }

    if (onProgress) {
      progressHandler = (message: IPCMessage) => {
        if (
          message.request_id === requestId &&
          message.session_id === sessionId &&
          message.type === 'sftp:progress'
        ) {
          onProgress(message.data as TransferProgress)
        }
      }
      backendService.on('sftp:progress', progressHandler)
    }

    const handler = (message: IPCMessage) => {
      if (message.request_id !== requestId) return

      if (message.session_id === sessionId && message.type === 'sftp:download') {
        cleanup(handler)
        resolve()
      } else if (message.type === 'error') {
        cleanup(handler)
        reject(new Error(message.data.error))
      }
    }

    backendService.on('sftp:download', handler)
    backendService.on('error', handler)

    requestId = backendService.send({
      type: 'sftp:download',
      session_id: sessionId,
      data: { remote_path: remotePath, local_path: localPath }
    })
  })
}

export const cancel = (sessionId: string, transferId: string): Promise<boolean> => {
  return new Promise((resolve, reject) => {
    let requestId = ''

    const handler = (message: IPCMessage) => {
      if (message.request_id !== requestId) return

      if (message.type === 'sftp:cancel') {
        backendService.off('sftp:cancel', handler)
        backendService.off('error', handler)
        resolve(message.data?.cancelled ?? false)
      } else if (message.type === 'error') {
        backendService.off('sftp:cancel', handler)
        backendService.off('error', handler)
        reject(new Error(message.data.error))
      }
    }

    backendService.on('sftp:cancel', handler)
    backendService.on('error', handler)

    requestId = backendService.send({
      type: 'sftp:cancel',
      session_id: sessionId,
      data: { transfer_id: transferId }
    })
  })
}

File: backend/internal/ipc/handlers/sftp/transfer.go
package sftp

import (
	"encoding/json"
	"fmt"
	"freessh-backend/internal/ipc/handlers"
	"freessh-backend/internal/models"
)

func (h *Handler) handleUpload(msg *models.IPCMessage, writer handlers.ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.UploadRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse upload request: %w", err)
	}

	progressChan := make(chan models.TransferProgress, 10)

	go func() {
		for progress := range progressChan {
			writer.WriteMessage(&models.IPCMessage{
				Type:      models.MsgSFTPProgress,
				SessionID: msg.SessionID,
				Data:      progress,
			})
		}
	}()

	err = h.manager.UploadFile(msg.SessionID, req.LocalPath, req.RemotePath, progressChan)
	close(progressChan)

	if err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPUpload,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "completed"},
	})
}

func (h *Handler) handleDownload(msg *models.IPCMessage, writer handlers.ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.DownloadRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse download request: %w", err)
	}

	progressChan := make(chan models.TransferProgress, 10)

	go func() {
		for progress := range progressChan {
			writer.WriteMessage(&models.IPCMessage{
				Type:      models.MsgSFTPProgress,
				SessionID: msg.SessionID,
				Data:      progress,
			})
		}
	}()

	err = h.manager.DownloadFile(msg.SessionID, req.RemotePath, req.LocalPath, progressChan)
	close(progressChan)

	if err != nil {
		return err
	}

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPDownload,
		SessionID: msg.SessionID,
		Data:      map[string]string{"status": "completed"},
	})
}

func (h *Handler) handleCancel(msg *models.IPCMessage, writer handlers.ResponseWriter) error {
	jsonData, err := json.Marshal(msg.Data)
	if err != nil {
		return fmt.Errorf("invalid data: %w", err)
	}

	var req models.CancelRequest
	if err := json.Unmarshal(jsonData, &req); err != nil {
		return fmt.Errorf("failed to parse cancel request: %w", err)
	}

	cancelled := h.manager.CancelTransfer(req.TransferID)

	return writer.WriteMessage(&models.IPCMessage{
		Type:      models.MsgSFTPCancel,
		SessionID: msg.SessionID,
		Data:      map[string]interface{}{"transfer_id": req.TransferID, "cancelled": cancelled},
	})
}
